server/src/main/java/com/thoughtworks/go/server/service/ScheduleService.java

```diff
@@ -728,6 +733,41 @@ public void failJob(JobInstance instance) {
         }
     }
 
+    /**
+     * IMPORTANT: this method is only meant for TOP level usage(never use this within a transaction). It gobbles exception.
+     */
+    public void rerunFailedJobs(String pipelineName, String counterOrLabel, String stageName, HttpOperationResult result) {
+        Pipeline pipeline = pipelineService.fullPipelineByCounterOrLabel(pipelineName, counterOrLabel);
+        if (pipelineDoesNotExist(pipelineName, counterOrLabel, result, pipeline)) return;
```

Comment:
return a 404 instead?

Reply:
pipelineDoesNotExist() and stageDoesNotExist() both return 404, it's check that is being used in two places, thats why I created a method.